enhance test framework to read meson arguments from a file per test

nioncode opened the pull request:
A `test_args.txt` file in the same directory as the test case will be parsed by the test framework and the content will be passed as arguments to meson during configuration.
The arguments are put before any `extra_args` to make them overwritable
from the command line.

This allows to test that `-D options` command line arguments properly overwrite default options and also allows every other command line argument to be tested for effectiveness, which might come in handy when implementing #429.

jpakkane commented:
I'm poorly on the net this weekend so I can't review. I'll get back to this probably on Monday.

Thanks,

rhd commented:
How does this interact with `mesonconf`? I think that letting another file in a test directory dictate build flags would lead to more confusion... but maybe helpful in some edge cases?

nioncode commented:
This file is only parsed from the meson _internal_ test framework that runs the test cases of meson itself.
It does *not* interact with any user facing tools like `mesonconf` or the initial `meson` configuration run at all.

jpakkane commented:
Looks like a useful addition, thanks. There are two minor fixes, thoug. First: that rather than having yet another custom format, let's instead use the same format as everywhere else, an array of strings. The sample file in question should look like this:

    # some comment here.
    ['--werror', '-D', 'testoption=A string with spaces', '-D', 'other_one=true', \
     '-D', 'combo_opt=one']

For parsing you might want to use `stringlistify` and the secret hidden (but slightly dirty) trick that you can find at `environment.py:669`.

Second: use `assert` instead of an if branch, it is cleaner. (Yes, not all of the existing code has been converted to use asserts, need to fix that at some point.)

nioncode commented:
Something like this?
All lines that are not empty and don't start with a '#' are concatenated and run through `eval`.
This is pretty unsafe though (might be fine because we only run this during meson's test cases?) and does not restrict the code that can run in any way.

nioncode commented:
Found a better way using [ast.literal_eval](https://docs.python.org/2/library/ast.html#ast.literal_eval), which only allows

> strings, numbers, tuples, lists, dicts, booleans, and `None`

It also takes care of any comments, but raises an Exception if no statement is given (i.e. empty or comment-only file). This should be fine for our testsuite, since we only provide the `test_args` file if we want to customize the options.

I squashed everything into one commit and rebased onto master :)

jpakkane commented:
Please fix the assert thing. We'll merge this after the next release (hopefully today).

Thanks,

nioncode commented:
Sorry, I thought you were talking about assert methods in the Python code, I didn't think about the test file.
Applied and squashed the change.
